### ISSUE TITLE
N significant digits rather than precision past decimal point?

### ISSUE THREAD

**Issue by mcskatkat** [Rockhopper-Technologies/prefixed#1]
This is well written and I've been trying it out for a while.
One thing that perhaps escapes me is whether it's possible to tweak the format specifiers to specify significant digits rather than fractional precision.

In many cases it makes scientific / engineering sense to produce N or up-to-N significant digits.
This follows the same rationale as the common scientific notation of M×10<sup>E</sup> with M given (up to) N digits past the decimal.
It also results in more predictable overall field width.

I've been trying to achieve this using the 'margin' format spec but I've come to conclude that it's probably impossible without code changes.
Is it?

Here are examples of existing output and desired result:

```
with          with 3           up-to 3
:.3h          sig. digits      sig. digits

200.000k       200k             200k
500.000k       500k             500k
900.000k       900k             900k
1.000M         1.00M            1M
1.500M         1.50M            1.5M
10.000M        10.0M            10M
15.000M        15.0M            15M
```

**Comment by avylove** [comment]
I can definitely see how that would be useful. I think it would require a couple new format specifiers.

One would work similarly to the decimal specifier. I'm not sure what would be good, but off the top of my head, if we used the asterix character (*), it would look something like `:*3h` and output `1.00M`.

We'd then need a flag to specify trailing 0s after the decimal should be dropped. Again, I'm not sure what this would be, but, for now, let's say the question mark (?), it would look something like `:?*3h` and output `1M`.

Is that what you're looking for?

I don't know when I'll be able to work on this. I don't have as much time for open source these days as I would like.

**Comment by mcskatkat** [comment]
I've made my own modifications to obtain a version for up-to-N-significant-digits. All it requires is using %g instead of %f when eventually formatting the unit-normalized float.

You are right saying that the problem is communicating the desired result through the format specifier. I'm not sure how to do this 'properly', i.e. in a non-confusing manner and without using up too many punctuation symbols.

My personal quick and dirty choice was the following, I believe it's possible to come up with something more coherent.
Also, this does not implement exactly-N-significant-digits, which is very useful when precision needs to be explicitly shown even for round numbers.


(only diffs shown below)
```
_SPEC_TYPE_MAP = dict(zip('hjJkKH', ('hf', 'jf', 'Jf', 'jg', 'Jg', 'hg')))
...
class Float(float):
    """...
    - Additional presentation types are supported for f-strings and :py:func:`format`.

      +---------+----------------------------------------------------------+
      | Type    | Meaning                                                  |
      +=========+==========================================================+
      | ``'h'`` | SI format. Outputs the number with closest divisible     |
      |         | SI prefix. (k, M, G, ...)                                |
      +---------+----------------------------------------------------------+
      | ``'j'`` | IEC Format. Outputs the number with closest divisible    |
      |         | IEC prefix. (Ki, Mi, Gi, ...)                            |
      +---------+----------------------------------------------------------+
      | ``'J'`` | Short IEC Format. Same as ``'j'`` but only a single      |
      |         | character.   (K, M, G, ...)                              |
      +---------+----------------------------------------------------------+
      | ``'k'`` | Same as ``'j'`` but precision specifies significant      |
      |         | digits rather than fractional precision                  |
      +---------+----------------------------------------------------------+
      | ``'K'`` | Same as ``'J'`` but precision specifies significant      |
      |         | digits rather than fractional precision                  |
      +---------+----------------------------------------------------------+
      | ``'H'`` | Same as ``'h'`` but precision specifies significant      |
      |         | digits rather than fractional precision                  |
      +---------+----------------------------------------------------------+
    ...
	"""
	...
    def __format__(self, format_spec):
		...
        spec_type = spec['type']

        spec_type = _SPEC_TYPE_MAP.get(spec_type)
        if spec_type is None:
            return super(Float, self).__format__(format_spec)
        spec_type, float_format_type = spec_type
		...
        if spec['precision'] is None:
            new_spec += float_format_type
        else:
            new_spec = '%s.%s%s' % (new_spec, spec['precision'], float_format_type)

        return '%s%s' % (value.__format__(new_spec), prefix)
```

**Comment by avylove** [comment]
I think your approach is probably more consistent with the Python spec. I think to get exactly-N-significant-digits, we'd add support for #, since that would match it's behavior with `g`/`G` where it disables removal of trailing zeros.

**Comment by mcskatkat** [comment]
Right, I forgot about `#` with `%g`. Thanks.
It works out of the box with my modification, just by using `#.3H` as format spec I get exactly-3-significant-digits. Great.

It's now up to you whether and how you want to include this in the package.

**Comment by avylove** [comment]
I think it would be a great addition, I just need to find time to incorporate it.

**Comment by avylove** [comment]
I started looking into implementing this, and I don't think using `g` directly is going to work. The problem is if you do something like `f'{Float(101):.2H}'` is going to output `1e+02`. I don't think exponent notation makes any sense since the library's purpose is to deal with magnitudes.

**Comment by mcskatkat** [comment]
Yes that is a quirk of `g` but it is caused by fewer significant digits than required to represent the integer part of the number.
I haven't encountered it because I tend to use a precision of 3, with which I'm not sure it can occur after scale normalization without using the 'margin' feature.

Still, it is a good question what needs to be displayed for 101 with 2 significant digits.
If you print 100, which uses 3 digits anyway, what's the point? Unless you interpret the number of significant digits to mean that the caller wants rounding. 

I don't find rounding intuitive and in my view, precision < 3 is not really a valid use case.
Nevertheless here's my workaround:

```python
from math import floor, log10
...
        if magnitude:
            ...
        else:
            ...

        if float_format_type == 'g':
            integer_digit_count = floor(log10(value))+1
            excess_digit_count = integer_digit_count - max(1, int(spec['precision']))
            if excess_digit_count > 0:
                value -= value % 10 ** excess_digit_count
                spec['precision'] = str(integer_digit_count)

        new_spec = ''.join(spec[key] for key in SPEC_FIELDS if spec[key] is not None)
        ...

```

**Comment by mcskatkat** [comment]
Actually the workaround above is too simplistic and also performs truncation instead of rounding.
A full solution requires more thought and some replication of the internal logic of the `g` format.

I'm still not sure that supporting fewer-than-integer digits is a required / reasonable use case.

**Comment by mcskatkat** [comment]
Here's a workaround that uses `%g` and post-processes it, instead of trying to outsmart it:

```python
        new_spec = ''.join(...)
        if spec['precision'] is None:
            ...
        else:
            ...

        formatted_value = value.__format__(new_spec)
        match = re.match(r'(\d+)(?:\.0*)?(?:[eE](\+\d+))?$', formatted_value)
        if match:
            formatted_value = str(int(match.group(1)) * 10**int(match.group(2) or 0))

        return formatted_value + prefix
```

Tests ok with this:
```python
for value in 1234.5, 999999:

    for i in range(8):
        fmt = f"{{:.{i}H}}"
        print("precision", i, fmt.format(Float(value)))
    print()

    for i in range(8):
        fmt = f"{{:#.{i}H}}"
        print("precision #", i, fmt.format(Float(value)))
    print()
```

**Comment by avylove** [comment]
The main problem with post-processing is you can't be sure what other formatters are in the spec. The user may be doing alignment or manipulating the sign.

I actually have a method where I pre-process that I'm pretty happy with except I wasn't able to determine if I have trailing zeros in the decimal without converting to a string. I was hoping to find a way to do it mathematically.

**Comment by avylove** [comment]
I have code that I'm happy with, just need to update the tests and make sure it all works for all the test cases. Sorry it's slow, but I just don't get the cycles to work on it that I want. One thing that came up was it seem a case like `f'{100.1:#.3H}'` will yield `100.`, with a trailing decimal. I don't like it, but it is consistent with the output of `f'{100.1:#.3g}'` and `f'{100.1:#.0f}'`.

**Comment by avylove** [comment]
@mcskatkat I have the code working and updated the tests. Still need to add a couple more tests and update the docs.
Decided to drop the trailing decimal, thinking the suffix is a good excuse for it.
I don't like how the cases mean different things, so thinking of changing the letters and leaving j/J as deprecated but there for backwards compatibility. Thinking k/K and m/M. That way the capital letter will always be the significant digits version.
Any thoughts on this?

**Comment by avylove** [comment]
Sorry this isn't moving faster. I put up my changes in https://github.com/Rockhopper-Technologies/prefixed/pull/2. I still need to add something to the docs about significant digits and the behavior of `#`, but the code should be good. @mcskatkat Can you take a look and let me know what you think?

**Comment by avylove** [comment]
Docs updated. I think this is ready to go. Will let it sit for a bit in case someone has comments, but expect to merge and release in the near future.

**Comment by avylove** [comment]
Released in [0.4.0](https://github.com/Rockhopper-Technologies/prefixed/releases/tag/0.4.0).
@mcskatkat Thanks for the suggestions!
